### PR TITLE
feat(appset): allow filtering repo by archived status

### PR DESCRIPTION
**Pull request by aburan28**
This PR adds a filter on repos to choose if archived repos are to be included in the scm generator.
For now this only includes github and gitea.

Closes [ISSUE #18531]

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by aburan28** [comment]
Hey @crenshaw-dev mind taking a look at this PR?

File: pkg/apis/application/v1alpha1/applicationset_types.go [hunk 1]
```diff
@@ -589,6 +589,8 @@ type SCMProviderGeneratorFilter struct {
 	LabelMatch *string `json:"labelMatch,omitempty" protobuf:"bytes,4,opt,name=labelMatch"`
 	// A regex which must match the branch name.
 	BranchMatch *string `json:"branchMatch,omitempty" protobuf:"bytes,5,opt,name=branchMatch"`
+	// Include archived repos.
+	IncludeArchivedRepos bool `json:"includeArchivedRepos,omitempty" protobuf:"varint,6,opt,name=includeArchivedRepos"`
```
Comment (rumstead) [inline pkg/apis/application/v1alpha1/applicationset_types.go:593]: I agree with @agaudreault's [comment](https://github.com/argoproj/argo-cd/issues/18531#issuecomment-2183576125). When using booleans and defaults, it is easier to have the default be false.

File: docs/operator-manual/applicationset/Generators-SCM-Provider.md [hunk 2]
```diff
@@ -410,6 +410,7 @@ spec:
 * `pathsDoNotExist`: An array of paths within the repository that must not exist. Can be a file or directory.
 * `labelMatch`: A regexp matched against repository labels. If any label matches, the repository is included.
 * `branchMatch`: A regexp matched against branch names.
+* `includeArchivedRepos`: Include repositories that are archived.
```
Comment (agaudreault) [inline docs/operator-manual/applicationset/Generators-SCM-Provider.md:413]: To be in the main config, it should be supported and implemented in all scm providers. I think this should be moved to the specific github and gitea if they are the only scm supporting it.

**Review by agaudreault — CHANGES_REQUESTED** [review]

File: docs/operator-manual/applicationset/Generators-SCM-Provider.md (hunk 2 above)
Comment (aburan28) [reply docs/operator-manual/applicationset/Generators-SCM-Provider.md:413]: Sounds good, I'll make that change.

File: assets/swagger.json [hunk 3]
```diff
@@ -8973,6 +8973,10 @@
           "description": "A regex which must match the branch name.",
           "type": "string"
         },
+        "includeArchivedRepos": {
```
Comment (agaudreault) [inline assets/swagger.json:8976]: Outdated. re-running codegen should fix the issues

File: pkg/apis/application/v1alpha1/applicationset_types.go [hunk 4]
```diff
@@ -506,6 +510,8 @@ type SCMProviderGeneratorGitlab struct {
 	Topic string `json:"topic,omitempty" protobuf:"bytes,8,opt,name=topic"`
 	// ConfigMap key holding the trusted certificates
 	CARef *ConfigMapKeyRef `json:"caRef,omitempty" protobuf:"bytes,9,opt,name=caRef"`
+	// Exclude repositories that are archived.
+	ExcludeArchivedRepos bool `json:"excludeArchivedRepos,omitempty" protobuf:"varint,6,opt,name=excludeArchivedRepos"`
```
Comment (agaudreault) [inline pkg/apis/application/v1alpha1/applicationset_types.go:514]: Added to gitlab struct, but I dont see modifications to services/scm_provider/gitlab.go

**Review by agaudreault — CHANGES_REQUESTED** [review]

File: docs/operator-manual/applicationset/Generators-SCM-Provider.md [hunk 5]
```diff
@@ -44,6 +44,8 @@ spec:
         api: https://git.example.com/
         # If true, scan every branch of every repository. If false, scan only the default branch. Defaults to false.
         allBranches: true
+        # Exclude repos that are archived
```
Comment (andrii-korotkov-verkada) [inline docs/operator-manual/applicationset/Generators-SCM-Provider.md:47]: Mention that it defaults to false similar to the comment above.

**Review by andrii-korotkov-verkada — CHANGES_REQUESTED** [review]
The deletion of generated pb files doesn't look intended. If anything, we should do this in a separate PR.

**Comment by andrii-korotkov-verkada** [comment]
The changes look relatively simple for each provider so far. How much effort would it be to add support for all providers that ArgoCD supports?

**Comment by aburan28** [comment]
> The deletion of generated pb files doesn't look intended. If anything, we should do this in a separate PR.

I am going to move this PR to a new PR to clean some things up here. https://github.com/argoproj/argo-cd/pull/20736

**Comment by aburan28** [comment]
> The changes look relatively simple for each provider so far. How much effort would it be to add support for all providers that ArgoCD supports?

Also I initially looked to this but not each provider has archived repos. 

For example bitbucket does not have this https://confluence.atlassian.com/bbkb/how-to-archive-repositories-1388151711.html
and gitlab has archived projects instead of repos.

**Pull request closed without merging**